### PR TITLE
Create OptimadeQueryWidget

### LINKED ISSUE #130: Implement AiiDAlab-specific OPTIMADE structure query and importer widget
Reuse the current widget from [`aiidalab/aiidalab-optimade`](https://github.com/aiidalab/aiidalab-optimade).

This was decided in a meeting between @yakutovicha and myself on 25.09.2020.

### PR DESCRIPTION
**Pull request by CasperWA**
Closes #130.

This widget is meant to mainly be used in AiiDAlab applications, it's specifically good in combination with the StructureManagerWidget.

### CODE REVIEW

**Review by csadorf — COMMENTED** [review]
Are we sure that this belongs into the widget-base package? I think it would be perfectly fine to leave it in aiidalab-optimade, no? One major issue I have with moving it into this package is, that it would require us to add "optimade" to the aiidalab-widgets-base dependencies.

**Comment by CasperWA** [comment]
> Are we sure that this belongs into the widget-base package? I think it would be perfectly fine to leave it in aiidalab-optimade, no? One major issue I have with moving it into this package is, that it would require us to add "optimade" to the aiidalab-widgets-base dependencies.

As per the discussion with @yakutovicha on 25.09.2020 this was the decided conclusion.

Let me try to give some reasons and motivations, but note this may not be complete:
@yakutovicha stressed that he would like to remove explicitly installing `optimade-client` or `aiidalab-optimade` in the docker stack.
This specific widget is not available in the `optimade-client` package.
We discussed moving it to the `optimade-client` package, but since we would have to do a `lambda` function transformation or similar each time one would use the widget in any case, to set the `structure` traitlet to either an `ase.Atoms` or `aiida.orm.StructureData`, we moved in another direction.
We discussed simply keeping things the way they were, but that would also mean importing some widgets from the `optimade-client` and the other from `aiidalab-optimade` - also not ideal.
I suggested simply adding the OPTIMADE `Structure` class to the `StructureManagerWidget`, but @yakutovicha was against this for maintenance reasons.
In the end, we settled on moving the only widget present in `aiidalab-optimade` to this repository and package, since the dependencies are not big to begin with, the use of this particular widget is AiiDAlab-specific, especially for use in the `StructureManagerWidget`, and if one wants to use it outside AiiDAlab, the widget is very simple to recreate from the widgets in `optimade-client`. In the end it's all MIT licensed, so a copy-paste of the widget we use here is no issue.

**Comment by csadorf** [comment]
> @yakutovicha stressed that he would like to remove explicitly installing `optimade-client` or `aiidalab-optimade` in the docker stack.

I think that makes a lot of sense.

> This specific widget is not available in the `optimade-client` package.

Sure.

> We discussed moving it to the `optimade-client` package, but since we would have to do a `lambda` function transformation or similar each time one would use the widget in any case, to set the `structure` traitlet to either an `ase.Atoms` or `aiida.orm.StructureData`, we moved in another direction.

I don't think that's an issue. Using small transformation functions is very common practice to resolve minor API friction.

> We discussed simply keeping things the way they were, but that would also mean importing some widgets from the `optimade-client` and the other from `aiidalab-optimade` - also not ideal.

Agreed, that would not be ideal. However I thought the widget was not in `optimade-client` to begin with? Furthermore, it makes total sense to me to import a widget that is designed to be used with AiiDAlab from the `aiidalab-optimade` package.

> I suggested simply adding the OPTIMADE `Structure` class to the `StructureManagerWidget`, but @yakutovicha was against this for maintenance reasons.

What maintenance burden exactly? I don't follow this argument.

> In the end, we settled on moving the only widget present in `aiidalab-optimade` to this repository and package, since the dependencies are not big to begin with, the use of this particular widget is AiiDAlab-specific, especially for use in the `StructureManagerWidget`, and if one wants to use it outside AiiDAlab, the widget is very simple to recreate from the widgets in `optimade-client`. In the end it's all MIT licensed, so a copy-paste of the widget we use here is no issue.

Is it not correct that `aiidalab-widgets-base` will now need to be extended by all requirements of `aiidalab-optimade`?

**Comment by CasperWA** [comment]
> > @yakutovicha stressed that he would like to remove explicitly installing `optimade-client` or `aiidalab-optimade` in the docker stack.
> 
> I think that makes a lot of sense.
> 
> > This specific widget is not available in the `optimade-client` package.
> 
> Sure.
> 
> > We discussed moving it to the `optimade-client` package, but since we would have to do a `lambda` function transformation or similar each time one would use the widget in any case, to set the `structure` traitlet to either an `ase.Atoms` or `aiida.orm.StructureData`, we moved in another direction.
> 
> I don't think that's an issue. Using small transformation functions is very common practice to resolve minor API friction.

Sure. But this would probably greatly reduce the ease with which it can be/is implemented in `StructureManagerWidget` or other similar widgets in AiiDAlab - but I partly agree with your assessment here.

> > We discussed simply keeping things the way they were, but that would also mean importing some widgets from the `optimade-client` and the other from `aiidalab-optimade` - also not ideal.
> 
> Agreed, that would not be ideal. However I thought the widget was not in `optimade-client` to begin with? Furthermore, it makes total sense to me to import a widget that is designed to be used with AiiDAlab from the `aiidalab-optimade` package.

Indeed. The widgets in `optimade-client` are the core widgets for this client. The widget in `aiidalab-optimade` is simply utilizing these widgets, making a combined "choose provider+database" and "set filters for searching" widget. Both of these "sub"-widgets are located in `optimade-client`.
The issue with keeping it in `aiidalab-optimade`, is that it violates the first problem - then we have to keep it in the docker stack, since `aiidalab-optimade` is not an installable PyPI package, and definitely doesn't need to be either.

> > I suggested simply adding the OPTIMADE `Structure` class to the `StructureManagerWidget`, but @yakutovicha was against this for maintenance reasons.
> 
> What maintenance burden exactly? I don't follow this argument.
>
> > In the end, we settled on moving the only widget present in `aiidalab-optimade` to this repository and package, since the dependencies are not big to begin with, the use of this particular widget is AiiDAlab-specific, especially for use in the `StructureManagerWidget`, and if one wants to use it outside AiiDAlab, the widget is very simple to recreate from the widgets in `optimade-client`. In the end it's all MIT licensed, so a copy-paste of the widget we use here is no issue.
> 
> Is it not correct that `aiidalab-widgets-base` will now need to be extended by all requirements of `aiidalab-optimade`?

Sorry, I meant the dependencies of this package (@yakutovicha was worried that if people wished to use this tiny widget outside AiiDAlab, they would suddenly have to rely on a major package - I don't see this as an issue for the reasons I have already stated).
The dependency of `aiidalab-optimade` is only `optimade-client`, this dependency was what we originally wanted to include in the list of docker stack requirements, i.e., installable packages from PyPI - not directly in the Dockerfile. Hence, the added dependency here would be 0 (effectively).

**Comment by yakutovicha** [comment]
> We discussed moving it to the `optimade-client` package, but since we would have to do a `lambda` function transformation or similar each time one would use the widget in any case, to set the `structure` traitlet to either an `ase.Atoms` or `aiida.orm.StructureData`, we moved in another direction.

I would like to comment on this. Essentially, if we will extend the `StructureManager`'s API to include a little transformation function, we could import the `OptimadeClient` that returns `OptimadeStructure` as is.

Currently, however, the `structure` trait of the widget is `dlink`ed to the `input_structure` trait:
https://github.com/aiidalab/aiidalab-widgets-base/blob/cce7876ab87b3b5876683c4a71c62dfc3117fbd9/aiidalab_widgets_base/structures.py#L128

**Comment by csadorf** [comment]
> > We discussed moving it to the `optimade-client` package, but since we would have to do a `lambda` function transformation or similar each time one would use the widget in any case, to set the `structure` traitlet to either an `ase.Atoms` or `aiida.orm.StructureData`, we moved in another direction.
> 
> I would like to comment on this. Essentially, if we will extend the `StructureManager`'s API to allow including a little transformation function, we could import the `OptimadeClient` that returns `OptimadeStructure` as is.

That's not necessary. It is very easily possible to add that transformation without API changes to `StructureManager`.

**Comment by yakutovicha** [comment]
> That's not necessary. It is very easily possible to add that transformation without API changes to `StructureManager`.

@csadorf can you make an example please?

**Comment by CasperWA** [comment]
> > > We discussed moving it to the `optimade-client` package, but since we would have to do a `lambda` function transformation or similar each time one would use the widget in any case, to set the `structure` traitlet to either an `ase.Atoms` or `aiida.orm.StructureData`, we moved in another direction.
> > 
> > 
> > I would like to comment on this. Essentially, if we will extend the `StructureManager`'s API to allow including a little transformation function, we could import the `OptimadeClient` that returns `OptimadeStructure` as is.
> 
> That's not necessary. It is very easily possible to add that transformation without API changes to `StructureManager`.

Indeed. However, I would probably go so far as to consider the possibility for a "catch-all" traitlet in the `StructureManagerWidget`, which can consume a large number of supported "Structure" classes, e.g., `ase.Atoms`, `aiida.orm.StructureData`, `aiida.orm.CifData`, `pymatgen.Structure`, `pymatgen.Molecule`, `optimade.adapters.Structure`, and have an internal transformation available for all these, but then only use one of these internally otherwise. This would make the `StructureManagerWidget` much more flexible in terms of developing widgets for it, and if specified correctly, the transformations should not be significant.

**Comment by csadorf** [comment]
> > That's not necessary. It is very easily possible to add that transformation without API changes to `StructureManager`.
> 
> @csadorf can you make an example please?

Sorry, you edited your comment after I had already replied. The simplest way I can think of for such a transformation under existing trait name is to simply write a thin wrapper:

```python

class FooWrapper(HBox):

    number = Int()
    def __init__(self, *args, **kwargs):
        self.foo = Foo(*args, **kwargs)
        dlink((foo, 'number'), (self, 'number'), transform=int)
        super().__init__(children=[self.foo])
```

**Comment by csadorf** [comment]
We might want to think a bit more about how best to enable the connection of importers with varying import formats. The kind of thin wrapper I proposed in my previous comment could be provided by a utility function that you only pass the transformation function or so. I'm not a fan of having "catch-all" traits that do internal transformations (something I've actively developed away from), because it makes the implementation of the manager very, very complex and difficult. But I see the problem and there is probably a way we can enable this a bit more intelligently without completely messing up the code internally.

**Comment by yakutovicha** [comment]
> The kind of thin wrapper I proposed in my previous comment could be provided by a utility function that you only pass the transformation function or so.

That what essentially what I was proposing here:

> Essentially, if we will extend the `StructureManager`'s API to allow including a little transformation function, we could import the `OptimadeClient` that returns `OptimadeStructure` as is.

If we know that the object returned by the Importer is different from (a) ASE Atoms and (b) AiiDA structure object, we can add a possibility to inject a little transformation. How exactly we do it is another question.

**Comment by CasperWA** [comment]
So it seems we are all talking for and against the same thing at the same time, but maybe with different scopes laid out in our heads (transformers for the `structure` traitlet).

**Comment by yakutovicha** [comment]
> So it seems we are all talking for and against the same thing at the same time, but maybe with different scopes laid out in our heads (transformers for the `structure` traitlet).

I feel we are converging :)

**Comment by yakutovicha** [comment]
Guys, what if we do something like:

widget = StructureManagerWidget(
    importers=[
        StructureUploadWidget(title="From computer"),
        CodQueryWidget(title="COD"),
        StructureBrowserWidget(title="AiiDA database"),
        SmilesWidget(title="SMILES"),  # requires OpenBabel!
        StructureExamplesWidget(
            title="From Examples",
            examples=[
                ("Silicon oxide", "miscellaneous/structures/SiO2.xyz")
            ]),
    ],
    editors = [
        BasicStructureEditor(title="Basic Editor"),
    ],
   transformators: {SmilesWidget: my_transform_function, ...}  # <<<--- The change.
)

**Comment by csadorf** [comment]
I'd prefer something like this:
```python
widget = StructureManagerWidget(
    importers=[
        StructureUploadWidget(title="From computer"),
        CodQueryWidget(title="COD"),
        StructureBrowserWidget(title="AiiDA database"),
        transform_structure(SmilesWidget(title="SMILES"), transform_function),
        StructureExamplesWidget(
            title="From Examples",
            examples=[
                ("Silicon oxide", "miscellaneous/structures/SiO2.xyz")
            ]),
    ],
    editors = [
        BasicStructureEditor(title="Basic Editor"),
    ],
)
```

**Comment by csadorf** [comment]
Actually, the best way would be to move the `dlink` out the structure manager constructor and into "user space". Because that is creating this issue in the first place.

**Comment by csadorf** [comment]
Happy to create a PR with revised API. If we are already breaking, then let's do it well. ;)

**Comment by yakutovicha** [comment]
> Happy to create a PR with revised API. If we are already breaking, then let's do a proper break. ;)

well, adding an optional parameter as I described won't break the API.

**Comment by csadorf** [comment]
> > Happy to create a PR with revised API. If we are already breaking, then let's do a proper break. ;)
> 
> well, adding an optional parameter as I described won't break the API.

My suggested alternative would also not break.

**Comment by CasperWA** [comment]
So the decision here is to implement a new API rather than merge the already present PR?

**Comment by csadorf** [comment]
> So the decision here is to implement a new API rather than merge the already present PR?

That would be my preferred option, but of course we need to find consensus.

**Comment by yakutovicha** [comment]
> > So the decision here is to implement a new API rather than merge the already present PR?
> 
> That would be my preferred option, but of course we need to find consensus.

I also agree with adapting API, but I am not sure about the right way do it. Let's maybe ask @giovannipizzi's input, when we meet for the AiiDAlab meeting.

**Comment by csadorf** [comment]
Proxy-based solution:
```python
from aiidalab_widgets_base import ImporterProxy

smiles_proxy = ImporterProxy()
dlink((SmilesWidget(title='SMILES'), 'some_structure_trait'), (smiles_proxy, 'structure'))

widget = StructureManagerWidget(
    importers=[
        StructureUploadWidget(title="From computer"),
        CodQueryWidget(title="COD"),
        StructureBrowserWidget(title="AiiDA database"),
        smiles_proxy,
        StructureExamplesWidget(
            title="From Examples",
            examples=[
                ("Silicon oxide", "miscellaneous/structures/SiO2.xyz")
            ]),
    ],
    editors = [
        BasicStructureEditor(title="Basic Editor"),
    ],
)
```

**Comment by CasperWA** [comment]
Final question, specific to this PR. Do we want to rename the widget, e.g., to `OptimadeStructuresWidget` or `AiidalabOptimadeWidget`? The current name was based on the existing `CodQueryWidget` name. To match this, but maybe it doesn't have to?

**Comment by yakutovicha** [comment]
> Final question, specific to this PR. Do we want to rename the widget, e.g., to `OptimadeStructuresWidget` or `AiidalabOptimadeWidget`? The current name was based on the existing `CodQueryWidget` name. To match this, but maybe it doesn't have to?

For me `AiidalabOptimadeWidget` is preferable.

File: aiidalab_widgets_base/structures.py [hunk 1]
```diff
@@ -8,17 +8,22 @@
 import ipywidgets as ipw
 from traitlets import Instance, Int, List, Unicode, Union, dlink, link, default, observe
 
-from sklearn.decomposition import PCA
+from optimade_client.query_filter import OptimadeQueryFilterWidget
+from optimade_client.query_provider import OptimadeQueryProviderWidget
```
Comment (CasperWA) [inline aiidalab_widgets_base/structures.py:12]: I could, in theory, move these imports down into the widget itself. They would then be imported _only_ together with the widget, but they would also be imported _every time_ with the widget ...

**Comment by yakutovicha** [comment]
As the conclusion of the meeting with @CasperWA, @csadorf, @giovannipizzi, @dou-du and myself we will keep the current implementation of the widget.

**Comment by CasperWA** [comment]
Ah, also. I just noticed the `CodQueryWidget` is actually in the module `databases`. Maybe I should move the OPTIMADE widget there as well?

File: aiidalab_widgets_base/structures.py (hunk 1 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/structures.py:12]: It is fine like this I think. We typically move all the imports to the top.

**Comment by CasperWA** [comment]
> > Final question, specific to this PR. Do we want to rename the widget, e.g., to `OptimadeStructuresWidget` or `AiidalabOptimadeWidget`? The current name was based on the existing `CodQueryWidget` name. To match this, but maybe it doesn't have to?
> 
> For me `AiidalabOptimadeWidget` is preferable.

Should we then also consider renaming `CodQueryWidget` to `AiidalabCodWidget`?

**Comment by yakutovicha** [comment]
> Ah, also. I just noticed the `CodQueryWidget` is actually in the module `databases`. Maybe I should move the OPTIMADE widget there as well?

yes, I would say so.

File: aiidalab_widgets_base/structures.py [hunk 2]
```diff
@@ -872,3 +877,49 @@ def remove(self, _):
         atoms = self.structure.copy()
         del [atoms[self.selection]]
         self.structure = atoms
+
+
+class OptimadeQueryWidget(ipw.VBox):
```
Comment (csadorf) [inline aiidalab_widgets_base/structures.py:882]: I would prefer to have the widget definition in a distinct module.

**Review by csadorf — CHANGES_REQUESTED** [review]

File: aiidalab_widgets_base/structures.py (hunk 2 above)
Comment (csadorf) [reply aiidalab_widgets_base/structures.py:882]: But this way is also fine since this is how most widgets are defined right now.

**Comment by yakutovicha** [comment]
> > > Final question, specific to this PR. Do we want to rename the widget, e.g., to `OptimadeStructuresWidget` or `AiidalabOptimadeWidget`? The current name was based on the existing `CodQueryWidget` name. To match this, but maybe it doesn't have to?
> > 
> > 
> > For me `AiidalabOptimadeWidget` is preferable.

Thinking more about it... I would like to withdraw my opinion about the widget's name. We have no widgets that start with `Aiidalab...`, so maybe `AiidalabOptimadeWidget` is still fine. Maybe we should agree on the naming policy as well.

> 
> Should we then also consider renaming `CodQueryWidget` to `AiidalabCodWidget`?

I would avoid any renamings for the moment, I don't want to deal with the compatibility issues now.

File: aiidalab_widgets_base/structures.py (hunk 2 above)
Comment (yakutovicha) [reply aiidalab_widgets_base/structures.py:882]: maybe move to the `databases` module as @CasperWA mentioned earlier.

Comment (CasperWA) [reply aiidalab_widgets_base/structures.py]: As opposed in the general `databases` file, where only `CodQueryWidget` is now?

**Comment by yakutovicha** [comment]
@CasperWA there is just a little pre-commit issue. Please fix it and I will approve.

**Review by csadorf — APPROVED** [review]
Approved pending that the style issue is fixed.

**Comment by CasperWA** [comment]
> @CasperWA there is just a little pre-commit issue. Please fix it and I will approve.

Them's be the issues of working on different computers throughout the week...
Fix is pushed.

On another note, I am not fond of the yapf styling. The "fix" is a rather ugly extra indentation. Also, we already moved away from `prospector` in `aiida-core` due to it being slow and causing some problems. I think mainly with its formatting choices and dependencies. Perhaps we should move to `black` instead?

**Review by yakutovicha — APPROVED** [review]

**Comment by yakutovicha** [comment]
Thanks @CasperWA

**Pull request merged**